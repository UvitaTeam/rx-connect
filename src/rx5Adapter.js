import Rx from "rxjs";
import ObservableSymbol from "symbol-observable";

export default {

    description: "RX5",

    Rx,

    next(o, value) {
        o.next(value);
    },

    isObservable(obj) {
        return obj && typeof obj[ObservableSymbol] === 'function';
    },

    unsubscribe(subscription) {
        subscription.unsubscribe();
    }
}
