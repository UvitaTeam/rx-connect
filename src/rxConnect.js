import React from "react";
import isPlainObject from "lodash.isplainobject";
import Rx from "./observable";
import ObservableSymbol from "symbol-observable";

const DEFAULT_OPTIONS = {
    noDebounce: false
};

function getDisplayName(WrappedComponent) {
    return WrappedComponent.displayName || WrappedComponent.name || "Component";
}
function isObservable(obj) {
    if (!obj) {
        return false;
    }

    if (Rx.Observable.isObservable) {
        return Rx.Observable.isObservable(obj);
    }

    return typeof obj[ObservableSymbol] === 'function';
}

export default function rxConnect(selector, options = DEFAULT_OPTIONS) {
    return WrappedComponent => class RxConnector extends React.PureComponent {

        static displayName = "RxConnector";

        stateSubscription = undefined;

        shouldDebounce = false;

        subProps = {};

        constructor(props) {
            super(props);

            this.props$ = new Rx.BehaviorSubject(props);
        }

        componentWillMount() {
            this.shouldDebounce = false;

            let mutations$ = selector;
            if(!isObservable(mutations$)) {
                mutations$ = selector(this.props$);
            }

            if(!isObservable(mutations$)) {
                    // eslint-disable-next-line no-undef
                if (typeof mutations$[Symbol.iterator] === 'function') {
                    mutations$ = Rx.Observable.merge(...mutations$);
                } else {
                    // eslint-disable-next-line no-console
                    console.error(`Selector must return an Observable, an iterator of observables. Got: `, mutations$);
                    return;
                }
            }

            this.stateSubscription = mutations$
                .scan((state, mutation) => {
                    if (isPlainObject(mutation)) {
                        return Object.assign({}, state, mutation);
                    }

                    if (typeof mutation === "function") {
                        return mutation(state);
                    }

                    // eslint-disable-next-line no-console
                    console.error(`Mutation must be a plain object or function. Check rxConnect of ${getDisplayName(WrappedComponent)}. Got: `, mutation);
                    return state;
                }, {})
                .debounce(() => (!options.noDebounce && this.shouldDebounce) ? Rx.Observable.interval(0) : Rx.Observable.of())
                .subscribe(state => {
                    this.subProps = state;
                    this.forceUpdate();
                });
        }

        componentDidMount() {
            this.shouldDebounce = true;
        }

        componentWillUnmount() {
            this.stateSubscription.dispose();
        }

        componentWillReceiveProps(nextProps) {
            if (this.props$.onNext) {
                this.props$.onNext(nextProps);
            } else {
                this.props$.next(nextProps);
            }
        }

        render() {
            return React.createElement(WrappedComponent, this.subProps, this.props.children);
        }
    };
}
