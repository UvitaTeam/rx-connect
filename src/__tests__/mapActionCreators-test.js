import { mapActionCreators, rxConnect } from "../";
import rx5Adapter from "../rx5Adapter";
import { getAdapter } from "../rxConnect";

const suites = {
    "RxJS 4": () => {},
    "RxJS 5": () => rxConnect.adapter = rx5Adapter
}

Object.entries(suites).forEach(([ name, initializer ]) => describe(name, () => {
    let Rx;
    beforeEach(() => {
        initializer();
        const adapter = getAdapter();
        Rx = adapter.Rx;
    });

    it("passes non-observables as is", async () => {
        const props = await mapActionCreators({ a: 123, b: "hi!" }).toPromise();

        expect(props).toMatchSnapshot();
    });

    it("strips dollar sign from Observable property names", async () => {
        const actions = {
            a$: new Rx.Subject()
        };

        const props = await mapActionCreators(actions).toPromise();

        expect(props).toMatchSnapshot();
    });

    it("creates FuncSubject-like action", async () => {
        const actions = {
            a$: new Rx.BehaviorSubject()
        };

        const props = await mapActionCreators(actions).toPromise();

        props.a(1, 2, 3);

        expect(actions.a$.getValue()).toMatchSnapshot();
    });

}));
