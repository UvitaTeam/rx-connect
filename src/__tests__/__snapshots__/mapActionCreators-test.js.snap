exports[`test creates FuncSubject-like action 1`] = `
Array [
  1,
  2,
  3,
]
`;

exports[`test passes non-observables as is 1`] = `
Object {
  "a": 123,
  "b": "hi!",
}
`;

exports[`test strips dollar sign from Observable property names 1`] = `
Object {
  "a": [Function],
}
`;
